List Data Type tutorial

avimallu opened the pull request:
Closes #330.

Open to feedback. No rust examples since I don't know rust, and can't find (from the docs) on how to (easily) manipulate strings in Rust.

avimallu commented:
@c-peters, should be okay. I don't know enough rust to add it here, unfortunately, and I can't find other examples in the docs in the string section for calling rust. 

The test-node failure seems unrelated, since I didn't touch that code.

c-peters commented on docs/user-guide/expressions/lists.md:
```diff
@@ -1,47 +1,102 @@
+Note: this is different from Python's `List` object, where the elements can be of any time. Polars can store these within columns, but as a generic `Object` datatype that doesn't have the special list manipulation features that we're about to discuss.
 
-This context is ideal for computing things in row orientation.
+# Powerful *List* manipulation
```
Please only use one H1 (# in markdown) for the title and use H2.... (##, ###, ...) for subtitles

c-peters commented on docs/user-guide/expressions/lists.md:
````diff
@@ -1,47 +1,102 @@
+--8<-- "python/user-guide/expressions/lists.py:weather_by_day"
+```
+
+Let's do something interesting. `pandas` allows you to compute the percentages of the rank values. `Polars` doesn't provide such a keyword argument. But because expressions are so versatile we can create our own percentage rank expression for highest temperature. Let's try that!
````
Could you remove the comparison with pandas. Users who are not familiar with the package or which are coming from other languages don't understand the reference

avimallu replied:
This is unchanged from the content (save for formatting) in the guide [currently present online;](https://pola-rs.github.io/polars-book/user-guide/expressions/lists/) and it is a good one. Are you sure you want it to go?

c-peters commented on docs/user-guide/expressions/lists.md:
```diff
@@ -1,47 +1,102 @@
+
+Polars provides several standard operations on `List` columns. If we want the first three measurements, we can do a `head(3)`. The last three can be obtained via a `tail(3)`, or alternately, via `slice` (negative indexing is supported). We can also identify the number of observations via `lengths`. Let's see them in action:
+
+{{code_block('user-guide/expressions/lists','list_ops',['list_ops'])}}
```
Use the third argument to highlight native functions you are using in the code snippet. You can add the function in API_REFERENCE_LINKS.yaml to create links to the documentation. If you only have python examples than only add the links for python

avimallu replied:
Ah, this is why my first attempts to do so were not working. Thank you very much!

c-peters commented on docs/user-guide/expressions/lists.md:
````diff
@@ -1,47 +1,102 @@
+--8<-- "python/user-guide/expressions/lists.py:list_ops"
 ```
 
+You can check out all the operations that are available by looking through the [`List` API documentation](https://pola-rs.github.io/polars/py-polars/html/reference/expressions/list.html). If you find references to the `arr` API on Stackoverflow or other sources, just replace `arr` with `List`, this was the old accessor for the `List` datatype.
````
See above use the API_REFERENCE_LINKS to supply links to the API documentation. This way we ensure it works for all languages.

c-peters replied:
Could you add the 'If you find references to the 'arr' sentence in an admonitions (https://squidfunk.github.io/mkdocs-material/reference/admonitions/#type:warning)?

c-peters commented on docs/user-guide/expressions/lists.md:
```diff
@@ -1,47 +1,102 @@
 
+The third step requires a casting (or alternately, a regex pattern search) operation to be perform on each element of the list. We can do this using by applying the operation on each element by first referencing them in the `pl.element()` context, and then calling a suitable Polars expression on them. Let's see how:
+
+{{code_block('user-guide/expressions/lists','count_errors',['count_errors'])}}
```
See above for third argument

c-peters commented on docs/user-guide/expressions/lists.md:
```diff
@@ -1,47 +1,102 @@
+
+What if we chose the regex route (i.e. recognizing the presence of *any* alphabetical character?)
+
+{{code_block('user-guide/expressions/lists','count_errors_regex',['count_errors_regex'])}}
```
See above for third argument

c-peters commented on docs/user-guide/expressions/lists.md:
````diff
@@ -1,47 +1,102 @@
 ```
 
-## Running polars expression on list elements
+If you're unfamiliar with the `(?i)`, it's a good time to look at the [`str.contains`](https://pola-rs.github.io/polars/py-polars/html/reference/expressions/api/polars.Expr.str.contains.html#polars-expr-str-contains) documentation for Polars! The rust regex crate provides a lot of additional regex flags that might come in handy.
````
See above for referencing the api documentation

c-peters commented on docs/user-guide/expressions/lists.md:
```diff
@@ -1,47 +1,102 @@
+
+This context is ideal for computing in row orientation.
+
+Polars expressions work on columns that have the guarantee that they consist of homogeneous data. Columns have this guarantee, rows in a `DataFrame` not so much. Luckily we have a data type that has the guarantee that the rows are homogeneous: `pl.List` data type.  We can **any** Polars operations on the elements of the list with the `list.eval` (`list().eval` in Rust) expression! These expressio ...
```
I'm not really following this paragraph, could you elaborate / rephrase?

avimallu replied:
This is from the current version itself. I agree it isn't the clearest as well, glad to know that we agree on it not being the best.

c-peters commented on docs/user-guide/expressions/lists.md:
```diff
@@ -1,47 +1,102 @@
-We can run **any** polars expression on the elements of a list with the `list.eval` (`list().eval` in Rust) expression! These expressions run entirely on polars' query engine and can run in parallel so will be super fast.
+Let's say we have another set of weather data across three days, for different stations:
+
+{{code_block('user-guide/expressions/lists','weather_by_day',['weather_by_day'])}}
```
See above for third argument

c-peters commented on docs/user-guide/expressions/lists.md:
```diff
@@ -1,47 +1,102 @@
-{{code_block('user-guide/expressions/lists','expression',['list.eval'])}}
+We can define `Array` columns in this manner:
 
+{{code_block('user-guide/expressions/lists','array_df',['array_df'])}}
```
See above for third argument

c-peters commented on docs/user-guide/expressions/lists.md:
````diff
@@ -1,47 +1,102 @@
 ```
 
-Note that this solution works for any expressions/operation you want to do row wise.
+Basic operations are available on it (see Array for details).
````
Could you provide a basic code snippet showing the functionality and then linking the user to the api documentation using the third argument and API_REFERENCE_LINKS

c-peters commented:
Hi @avimallu , I reviewed your PR. Thanks for the contribution, the comments mostly refer to how to link to the API documentation and to ensure the documentation remains agnostic to languages other than Python. That way we can insert node js / rust examples without changes the main documenation.

avimallu commented:
@c-peters, thank you so much for patiently pointing out where the pull request was lacking. I've addressed all your comments; just wanted to make sure you wanted to remove the `pandas` comparison for percentage ranks - it was a good one that's currently present and shows the expressiveness of the expression API.

c-peters commented:
LGTM, you can leave the Pandas paragraph in then.

avimallu commented:
Awesome! Ready to merge from my end.
